src/libraries/System.Private.CoreLib/src/System/Reflection/InvokeSignatureInfo.cs

```diff
@@ -0,0 +1,207 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Reflection
+{
+    internal sealed class InvokeSignatureInfo
+    {
+        internal readonly Type? _declaringType;
+        internal readonly Type[] _parameterTypes;
+        internal readonly Type _returnType;
+        internal readonly bool _isStatic;
+
+        public static InvokeSignatureInfo Create(in InvokeSignatureInfoKey key)
+        {
+            return new InvokeSignatureInfo(
+                key._declaringType,
+                key._parameterTypes,
+                key._returnType,
+                key._isStatic);
+        }
+
+        public InvokeSignatureInfo(Type? declaringType, Type[] parameterTypes, Type returnType, bool isStatic)
+        {
+            _declaringType = declaringType;
+            _parameterTypes = parameterTypes;
+            _returnType = returnType;
+            _isStatic = isStatic;
+        }
+
+        // Must be the same as InvokeSignatureInfoKey.Comparer.Equals().
+        public override bool Equals(object? other)
+        {
+            if (ReferenceEquals(this, other))
+            {
+                return true;
+            }
+
+            if (other is not InvokeSignatureInfo otherSig)
+            {
+                return false;
+            }
+
+            if (!ReferenceEquals(_declaringType, otherSig._declaringType) ||
+                !ReferenceEquals(_returnType, otherSig._returnType) ||
+                _isStatic != otherSig._isStatic ||
+                _parameterTypes.Length != otherSig._parameterTypes.Length)
+            {
+                return false;
+            }
+
+            for (int i = 0; i < _parameterTypes.Length; i++)
+            {
+                if (!ReferenceEquals(_parameterTypes[i], otherSig._parameterTypes[i]))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        public override int GetHashCode() => GetHashCode(_declaringType, _parameterTypes, _returnType);
+
+        public static int GetHashCode(Type? declaringType, Type[] parameterTypes, Type returnType)
+        {
+            int hashcode = 0;
```

Comment:
I think this impl can be simplified by using [`HashCode`](https://learn.microsoft.com/dotnet/api/system.hashcode). The explicit use of `int.Rotate*` to construct a new hash seems unnecessary when we have utilities to combine and construct a well defined new hash.

Reply:
`HashCode` is randomized hashcode. For this one, I think we may be rather going towards non-randomized stable hashcode so that would allow us to create hashtable of precompiled signatures at build time.

Reply:
Ah. That makes sense. Perhaps a comment on that expectation is warranted then?

Reply:
Currently the hashcode is based on combining several calls to `System.Type.GetHashCode()`, so it is not stable. It would have to use `Type.FullName`. I can make that change here, but since it not needed yet I think I should take the original suggestion of using `HashCode.Combine()` -- I checked that perf against current (rotate \ xor) and it's a wash.

Reply:
Also with a `Type.FullName` hashcode, I'd have to find or implement a fast stable version of that since the default is randomized.